File: hse_learning/go/contest_Y/num_C.go
package main

import (
	"fmt"
	"sort"
)

type Taxi struct {
	lastTimestamp int64
	position      int
}

func main() {
	var N, L, S int
	if _, err := fmt.Scan(&N, &L, &S); err != nil {
		panic(err)
	}

	taxis := make(map[int]Taxi)
	orderResults := make([][]int, 0)

	for i := 0; i < N; i++ {
		var typ string
		var timestamp int64
		if _, err := fmt.Scan(&typ, &timestamp); err != nil {
			panic(err)
		}

		if typ == "TAXI" {
			var taxiId, position int
			if _, err := fmt.Scan(&taxiId, &position); err != nil {
				panic(err)
			}
			taxis[taxiId] = Taxi{lastTimestamp: timestamp, position: position}
		} else if typ == "ORDER" {
			var orderId int
			var orderPosition, orderTime int
			if _, err := fmt.Scan(&orderId, &orderPosition, &orderTime); err != nil {
				panic(err)
			}
			_ = orderId // чтобы избежать ошибки «declared and not used»

			availableTaxis := make([]int, 0)
			for taxiId, taxi := range taxis {
				timeDiff := timestamp - taxi.lastTimestamp
				maxDistance := timeDiff * int64(S)
				minPos := taxi.position
				maxPos := (taxi.position + int(maxDistance)) % L

				var isIn bool
				if minPos <= maxPos {
					isIn = minPos <= orderPosition && orderPosition < maxPos
				} else {
					isIn = orderPosition >= minPos || orderPosition < maxPos
				}

				var worstPosition int
				if maxDistance >= int64(L) {
					worstPosition = (orderPosition + 1) % L
				} else {
					if isIn {
						worstPosition = (orderPosition + 1) % L
					} else {
						worstPosition = minPos
					}
				}

				distance := (orderPosition - worstPosition + L) % L

				if distance <= orderTime*S {
					availableTaxis = append(availableTaxis, taxiId)
				}
			}

			if len(availableTaxis) == 0 {
				orderResults = append(orderResults, []int{-1})
			} else {
				sort.Ints(availableTaxis)
				if len(availableTaxis) > 5 {
					availableTaxis = availableTaxis[:5]
				}
				orderResults = append(orderResults, availableTaxis)
			}
		}
	}

	for _, result := range orderResults {
		for i, val := range result {
			if i > 0 {
				fmt.Print(" ")
			}
			fmt.Print(val)
		}
		fmt.Println()
	}
}

File: hse_learning/cpp_lang/contest_Y/num_C.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

struct Taxi {
    long long lastTimestamp;
    int position;
};

int main() {
    int N, L, S;
    cin >> N >> L >> S;

    map<int, Taxi> taxis;
    vector<vector<int>> orderResults;

    for (int i = 0; i < N; ++i) {
        string type;
        long long timestamp;
        cin >> type >> timestamp;

        if (type == "TAXI") {
            int taxiId, position;
            cin >> taxiId >> position;
            taxis[taxiId] = {timestamp, position};
        } else if (type == "ORDER") {
            int orderId, orderPosition, orderTime;
            cin >> orderId >> orderPosition >> orderTime;

            vector<int> availableTaxis;
            for (auto& [taxiId, taxi] : taxis) {
                long long timeDiff = timestamp - taxi.lastTimestamp;
                int maxDistance = timeDiff * S;
                int minPos = taxi.position;
                int maxPos = (taxi.position + maxDistance) % L;

                bool isIn;
                // проверка: заказ внутри или снаружи области таксиста
                // 1. не пересекает начало круга
                if (minPos <= maxPos)
                    isIn = minPos <= orderPosition && orderPosition < maxPos;
                // 2. пересекает начало круга
                else
                    isIn = orderPosition >= minPos || orderPosition < maxPos;

                int worstPosition;
                if (maxDistance >= L) {
                    worstPosition = (orderPosition + 1) % L;
                }
                else {
                    if (isIn)
                        worstPosition = (orderPosition + 1) % L;
                    else
                        worstPosition = minPos;
                }

                // TAXI <timestamp> <taxi_id> <taxi_position>
                // ORDER <timestamp> <order_id> <order_position> <order_time>
                int distance = (orderPosition - worstPosition + L) % L;

                if (distance <= orderTime * S) {
                    availableTaxis.push_back(taxiId);
                }
            }

            if (availableTaxis.empty()) {
                orderResults.push_back({-1});
            } else {
                sort(availableTaxis.begin(), availableTaxis.end());
                if (availableTaxis.size() > 5) {
                    availableTaxis.resize(5);
                }
                orderResults.push_back(availableTaxis);
            }
        }
    }

    for (const auto& result : orderResults) {
        for (size_t i = 0; i < result.size(); ++i) {
            cout << result[i];
            if (i + 1 < result.size()) {
                cout << " ";
            }
        }
        cout << endl;
    }

    return 0;
}